🐛 BUG: `minify` changes names of React components (`keepNames: true` fixes)

### Which Cloudflare product(s) does this pertain to?

Wrangler

### What version(s) of the tool(s) are you using?

3.80.0

### What version of Node are you using?

_No response_

### What operating system and version are you using?

Mac Sonoma 14.6.1

### Describe the Bug

### Observed behavior

With `minify = true` in my Next.js app with stream resuming, I see the following errors in my Worker:

```
✘ [ERROR] ⨯ Error: Expected the resume to render <n5> in this slot but instead it rendered <Il>. The tree doesn't match so React will fallback to client rendering.
```

This does not occur if I use `minify = false`.

### Explanation

This is because `esbuild` has mangled the names when building with minification, so the React component names in the Worker JS script used for server-side rendering no longer match the component names used when the static shell was generated at build time that React is trying to render with.

The specifics of this don't matter too much – it's more that esbuild's minification is pretty aggressive and this name mangling would probably affect more use cases than just this one – however, it can be relaxed with the use of [esbuild's `keepNames` option](https://esbuild.github.io/api/#keep-names).

If I manually modify wrangler's `cli.js` to add `keepNames: true` as an option to the `esbuild.build()` call in `bundleWorker()`, but keep `minify = true`, this error goes away.

So my suggestion to fix this issue would be to add `keepNames: true` to the esbuild options – or expose `keepNames` as a wrangler option (though this seems unnecessarily specific) – or expose any esbuild option to wrangler (I would love this, but I know there's coupling arguments against this)

### Example results

Here are some example results to show how much impact this has on minification:

`minify: false`:

```
Total Upload: 3653.57 KiB / gzip: 822.45 KiB
```

`minify: true, keepNames: false` (or unspecified):

```
Total Upload: 1834.57 KiB / gzip: 654.76 KiB
```

`minify: true, keepNames: true`:

```
Total Upload: 1888.91 KiB / gzip: 672.51 KiB
```

So it results in a 2.7% increase in size compared with full minification – but considering the (uncompressed) minified result is half of the unminified result, still well worth it for a reduction in parsing/compilation time at cold start.

My suggestion would just be to add `keepNames: true` regardless – but I'll leave that up to the team to mull over.


### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

_No response_

--- a/.changeset/fluffy-colts-give.md
+++ b/.changeset/fluffy-colts-give.md
@@ -1,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+fix: enable esbuild's keepNames: true to set .name on functions/classes

--- a/packages/wrangler/src/__tests__/middleware.test.ts
+++ b/packages/wrangler/src/__tests__/middleware.test.ts
@@ -870,7 +870,11 @@ describe("middleware", () => {
 					.replace(/\/\/ .*/g, "")
 					.trim()
 			).toMatchInlineSnapshot(`
-				"var src_default = {
+				"var __defProp = Object.defineProperty;
+				var __name = (target, value) => __defProp(target, \\"name\\", { value, configurable: true });
+
+
+				var src_default = {
 				  async fetch(request, env) {
 				    return Response.json(env);
 				  }
@@ -882,6 +886,7 @@ describe("middleware", () => {
 				    return new Response(\\"Hello World\\");
 				  }
 				};
+				__name(DurableObjectExample, \\"DurableObjectExample\\");
 
 
 				var __INTERNAL_WRANGLER_MIDDLEWARE__ = [
@@ -894,6 +899,7 @@ describe("middleware", () => {
 				function __facade_register__(...args) {
 				  __facade_middleware__.push(...args.flat());
 				}
+				__name(__facade_register__, \\"__facade_register__\\");
 				function __facade_invokeChain__(request, env, ctx, dispatch, middlewareChain) {
 				  const [head, ...tail] = middlewareChain;
 				  const middlewareCtx = {
@@ -904,12 +910,14 @@ describe("middleware", () => {
 				  };
 				  return head(request, env, ctx, middlewareCtx);
 				}
+				__name(__facade_invokeChain__, \\"__facade_invokeChain__\\");
 				function __facade_invoke__(request, env, ctx, dispatch, finalMiddleware) {
 				  return __facade_invokeChain__(request, env, ctx, dispatch, [
 				    ...__facade_middleware__,
 				    finalMiddleware
 				  ]);
 				}
+				__name(__facade_invoke__, \\"__facade_invoke__\\");
 
 
 				var __Facade_ScheduledController__ = class {
@@ -926,6 +934,7 @@ describe("middleware", () => {
 				    this.#noRetry();
 				  }
 				};
+				__name(__Facade_ScheduledController__, \\"__Facade_ScheduledController__\\");
 				function wrapExportedHandler(worker) {
 				  if (__INTERNAL_WRANGLER_MIDDLEWARE__ === void 0 || __INTERNAL_WRANGLER_MIDDLEWARE__.length === 0) {
 				    return worker;
@@ -933,16 +942,16 @@ describe("middleware", () => {
 				  for (const middleware of __INTERNAL_WRANGLER_MIDDLEWARE__) {
 				    __facade_register__(middleware);
 				  }
-				  const fetchDispatcher = function(request, env, ctx) {
+				  const fetchDispatcher = /* @__PURE__ */ __name(function(request, env, ctx) {
 				    if (worker.fetch === void 0) {
 				      throw new Error(\\"Handler does not export a fetch() function.\\");
 				    }
 				    return worker.fetch(request, env, ctx);
-				  };
+				  }, \\"fetchDispatcher\\");
 				  return {
 				    ...worker,
 				    fetch(request, env, ctx) {
-				      const dispatcher = function(type, init) {
+				      const dispatcher = /* @__PURE__ */ __name(function(type, init) {
 				        if (type === \\"scheduled\\" && worker.scheduled !== void 0) {
 				          const controller = new __Facade_ScheduledController__(
 				            Date.now(),
@@ -952,11 +961,12 @@ describe("middleware", () => {
 				          );
 				          return worker.scheduled(controller, env, ctx);
 				        }
-				      };
+				      }, \\"dispatcher\\");
 				      return __facade_invoke__(request, env, ctx, dispatcher, fetchDispatcher);
 				    }
 				  };
 				}
+				__name(wrapExportedHandler, \\"wrapExportedHandler\\");
 				function wrapWorkerEntrypoint(klass) {
 				  if (__INTERNAL_WRANGLER_MIDDLEWARE__ === void 0 || __INTERNAL_WRANGLER_MIDDLEWARE__.length === 0) {
 				    return klass;
@@ -995,6 +1005,7 @@ describe("middleware", () => {
 				    }
 				  };
 				}
+				__name(wrapWorkerEntrypoint, \\"wrapWorkerEntrypoint\\");
 				var WRAPPED_ENTRY;
 				if (typeof middleware_insertion_facade_default === \\"object\\") {
 				  WRAPPED_ENTRY = wrapExportedHandler(middleware_insertion_facade_default);

--- a/packages/wrangler/src/deployment-bundle/bundle.ts
+++ b/packages/wrangler/src/deployment-bundle/bundle.ts
@@ -384,6 +384,7 @@ export async function bundleWorker(
 		bundle,
 		absWorkingDir: entry.directory,
 		outdir: destination,
+		keepNames: true,
 		entryNames: entryName || path.parse(entryFile).name,
 		...(isOutfile
 			? {

